File: 0743-network-delay-time/0743-network-delay-time.js
/**
 * @param {number[][]} times
 * @param {number} n
 * @param {number} k
 * @return {number}
 */
const networkDelayTime = function(times, n, k) {
    const adj = new Map();

    for (let i = 1; i <= n; i++) adj.set(i, []);
    for (let [u, v, w] of times) {
        adj.get(u).push([v, w]);
    }

    const dist = Array(n+1).fill(Infinity);
    dist[k] = 0;

    const pq = [[0, k]];

    while (pq.length > 0) {
        pq.sort((a, b) => a[0] - b[0]);
        const [currTime, node] = pq.shift();

        if (currTime > dist[node]) continue;

        for (const [neighbor,time] of adj.get(node)) {
            const newTime = currTime + time;
            if (newTime < dist[neighbor]) {
                dist[neighbor] = newTime;
                pq.push([newTime, neighbor]);
            }
        }
    }
    const maxTime = Math.max(...dist.slice(1));
    return maxTime === Infinity ? -1 : maxTime;
};

File: 0743-network-delay-time/0743-network-delay-time.py
from collections import defaultdict
import heapq

class Solution(object):
    def networkDelayTime(self, times, n, k):
        """
        :type times: List[List[int]]
        :type n: int
        :type k: int
        :rtype: int
        """
        graph = defaultdict(list)
        distances = [float('inf')] * (n+1)
        distances[0], distances[k] = 0, 0

        for u, v, w in times:
            graph[u].append((v, w))

        queue = []
        heapq.heappush(queue, (0, k))   

        while queue:
            dist, node = heapq.heappop(queue)

            if dist < distances[node]:
                continue

            print(distances)    
            for neighbor, cost in graph[node]:
                new_cost = cost + dist
                if new_cost < distances[neighbor]:
                    distances[neighbor] = new_cost
                    heapq.heappush(queue, (new_cost, neighbor))

        return max(distances) if max(distances) != float('inf') else -1